File: api/internal/offer/offer.go
// Package offer is ADR-0008's Offer flow (#317): the event that grants a
// Doula's attachment to one Engagement through her own agreement. An
// Owner or Admin offers work to an existing doula Membership or to a bare
// email address, fan-out is uncapped and the first acceptance wins,
// acceptance mints a granted engagement_attachments row with the fee and
// terms copied onto it, and completion of the Engagement closes whatever
// is still open.
//
// An Offer is a copy, never a view: the four decidable facts it carries
// (Client first initial, general area, exact due date, fee) plus its
// free-text terms are typed in at send time and never refreshed, which
// is what lets someone who is not Staff at all read enough to decide
// before she has an account (#230). A changed fact means withdraw and
// re-offer.
package offer

import (
	"context"
	"crypto/rand"
	"database/sql"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"time"

	"doula-cloud/api/internal/staffauth"
)

// Lifetime is how long an Offer stays open after it is sent -- 7 days,
// the default #229 chose and ADR-0008 records, the same as an
// Invitation's.
const Lifetime = 7 * 24 * time.Hour

// maxAccessCodeAttempts bounds guessing at the pre-account read's
// six-digit code. The endpoint sits outside staffauth, so the code is the
// whole second factor in front of it; ten guesses against a 10^6 space is
// a 1-in-100,000 chance of a hit, and a burned Offer stays burned because
// the counter lives on the row (00041) rather than in a process.
const maxAccessCodeAttempts = 10

// newAccessCode returns the six-digit code the pre-account Offer read
// asks for, from crypto/rand rather than math/rand: it is a credential,
// short enough that a predictable one would be trivially reproducible.
// Formatted with leading zeros so every code is six characters -- a
// five-digit one would leak that the number is small.
func newAccessCode() (string, error) {
	n, err := rand.Int(rand.Reader, big.NewInt(1_000_000))
	if err != nil {
		// coverage:ignore reason: crypto/rand failure, not exercised by unit tests
		return "", fmt.Errorf("offer: generate access code: %w", err)
	}
	return fmt.Sprintf("%06d", n.Int64()), nil
}

// expireOpen flips every Offer matching the given engagement/staff/id
// filter that has run past its own expires_at to 'expired', on the way
// past. There is no sweep job: acceptInvite (staffauth) set the
// precedent, and for the same reason -- the person looking at the row is
// the one who found out it is stale, and every read and decision path in
// this package calls this first so no path can act on an Offer that has
// quietly outlived itself.
//
// column is one of the three this package filters Offers by, never
// caller input -- see the named constants below.
func expireOpen(ctx context.Context, tx *sql.Tx, column, value string) error {
	if _, err := tx.ExecContext(ctx,
		//nolint:gosec // column is one of three package-local constants, never request input
		`UPDATE engagement_offers SET state = 'expired'
		  WHERE `+column+` = $1 AND state = 'offered' AND expires_at <= now()`,
		value,
	); err != nil {
		// coverage:ignore reason: DB query failure, not exercised by unit tests
		return fmt.Errorf("offer: expire open offers: %w", err)
	}
	return nil
}

// The three columns expireOpen filters by, named so no call site can pass
// a string that came from a request.
const (
	byID           = "id"
	byEngagementID = "engagement_id"
	byStaffID      = "staff_id"
)

// The offer_state values (00030) this package compares against by name.
// The database is the authority on the set; these exist so a typo is a
// compile error rather than a condition that quietly never matches.
const (
	stateOffered  = "offered"
	stateDeclined = "declined"
)

// writeJSON encodes a 200 response body, the one shape every handler in
// this package shares.
func writeJSON(w http.ResponseWriter, body any) {
	w.Header().Set("Content-Type", "application/json")
	// coverage:ignore reason: response encoding failure, not exercised by unit tests
	if err := json.NewEncoder(w).Encode(body); err != nil {
		http.Error(w, staffauth.MsgInternalError, http.StatusInternalServerError)
	}
}

File: app/src/lib/offer.ts
/**
 * An Offer of one Engagement's work (ADR-0008, #317): an Owner or Admin
 * offers it to an existing Doula or to an email address, and the person
 * offered accepts or declines. Decoupled from SvelteKit and the DOM so it
 * can be unit-tested directly -- mirrors invoice.ts and contract.ts.
 *
 * The four decidable facts (Client first initial, general area, exact due
 * date, fee) plus free-text terms are the whole of what an Offer carries.
 * They are a copy taken at send time, never a live view of the Engagement,
 * which is what lets the same record be read by someone who has no
 * account yet.
 */

/** A minimal fetch-shaped function, injected rather than imported -- see
 * contract.ts's Fetcher for why. */
export type Fetcher = (path: string, init?: RequestInit) => Promise<Response>;

/** One Offer as either side reads it. targetName/targetAddress are filled
 * only on the Practice-side read: her own inbox does not need to tell her
 * who she is. */
export interface Offer {
	offerId: string;
	state: OfferState;
	/** The three fields #230 calls the Client's own. They come back empty
	 * once the Offer reaches a terminal state: the record of the asking
	 * survives, the Client's details stop being served. */
	clientFirstInitial: string;
	clientArea: string;
	dueDate: string;
	amountCents?: number;
	terms?: string;
	employmentType: string;
	offeredAt: string;
	expiresAt: string;
	decidedAt?: string;
	targetName?: string;
	targetAddress?: string;
}

/** The six states an Offer's lifecycle passes through -- the offer_state
 * enum (00030), mirrored so a typo is a type error. */
export type OfferState = 'offered' | 'accepted' | 'declined' | 'withdrawn' | 'superseded' | 'expired';

/** What the pre-account read serves: the same decidable facts, and
 * nothing that would name the Client, the Practice, or the Engagement. */
export interface PreAccountOffer {
	offerId: string;
	state: OfferState;
	clientFirstInitial: string;
	clientArea: string;
	dueDate: string;
	amountCents?: number;
	terms?: string;
	expiresAt: string;
}

/** The body of a make-an-offer request. Exactly one of staffId and email
 * is set. Employment type is never sent: it is read off her Membership
 * for a staffId target, and an emailed Invitation always joins her as a
 * contractor, which is why that path always carries a fee. */
export interface NewOffer {
	staffId?: string;
	email?: string;
	amountCents?: number;
	terms?: string;
	clientFirstInitial: string;
	clientArea: string;
	dueDate: string;
}

/** Which Badge variant each state wears -- kept beside the labels so the
 * two never drift, and so neither screen owns a copy of the other's. */
export const offerStateVariants: Record<OfferState, 'info' | 'success' | 'warning' | 'neutral'> = {
	offered: 'info',
	accepted: 'success',
	declined: 'neutral',
	withdrawn: 'neutral',
	superseded: 'neutral',
	expired: 'warning'
};

/** Human-readable labels for each state, so a list reads as a story
 * rather than as enum values. */
export const offerStateLabels: Record<OfferState, string> = {
	offered: 'Awaiting a decision',
	accepted: 'Accepted',
	declined: 'Declined',
	withdrawn: 'Withdrawn',
	superseded: 'Taken by someone else',
	expired: 'Expired'
};

function engagementOffersPath(practiceId: string, engagementId: string): string {
	return `/api/practices/${practiceId}/engagements/${engagementId}/offers`;
}

function offerActionPath(practiceId: string, offerId: string, action: string): string {
	return `/api/practices/${practiceId}/offers/${offerId}/${action}`;
}

async function readOrThrow<T>(response: Response): Promise<T> {
	if (!response.ok) {
		throw new Error(await response.text());
	}
	return response.json() as Promise<T>;
}

/** Loads every Offer made on one Engagement, newest first -- the
 * make-an-offer screen's own list of who has been asked and what each of
 * them said. Owner and Admin only. Only the first page: the BFF paginates
 * for the reason docs/api-design.md gives, not because a fan-out ever
 * reaches a second page (see invoice.ts's loadInvoices). */
export async function loadEngagementOffers(
	fetcher: Fetcher,
	practiceId: string,
	engagementId: string
): Promise<Offer[]> {
	const body = await readOrThrow<{ items: Offer[] }>(
		await fetcher(engagementOffersPath(practiceId, engagementId))
	);
	return body.items;
}

/**
 * Loads the caller's own Offers, open and past -- her inbox. First page
 * only, same as loadEngagementOffers.
 */
export async function loadInbox(fetcher: Fetcher, practiceId: string): Promise<Offer[]> {
	const body = await readOrThrow<{ items: Offer[] }>(await fetcher(`/api/practices/${practiceId}/offers`));
	return body.items;
}

/** Makes an Offer. Throws with the response body text on refusal -- a
 * missing fee for a contractor, an address that already holds a
 * membership, a target who is not a Doula. */
export async function createOffer(
	fetcher: Fetcher,
	practiceId: string,
	engagementId: string,
	offer: NewOffer
): Promise<{ offerId: string; expiresAt: string }> {
	return readOrThrow(
		await fetcher(engagementOffersPath(practiceId, engagementId), {
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify(offer)
		})
	);
}

/** Decides one of the caller's own Offers: 'accept' or 'decline'.
 * Accepting is what mints her attachment to the Engagement. */
export async function decideOffer(
	fetcher: Fetcher,
	practiceId: string,
	offerId: string,
	action: 'accept' | 'decline'
): Promise<{ offerId: string; state: OfferState }> {
	return readOrThrow(await fetcher(offerActionPath(practiceId, offerId, action), { method: 'POST' }));
}

/** Takes an Offer back. The Practice's answer to a fact that moved: an
 * Offer is never edited after it is sent, so this is half of
 * withdraw-and-re-offer. */
export async function withdrawOffer(
	fetcher: Fetcher,
	practiceId: string,
	offerId: string
): Promise<{ offerId: string; state: OfferState }> {
	return readOrThrow(await fetcher(offerActionPath(practiceId, offerId, 'withdraw'), { method: 'POST' }));
}

/** Reads an Offer with the emailed link's token and the six-digit code,
 * before the reader has any account. Uses a bare fetcher with no session:
 * the two credentials are the whole of the authentication. */
export async function loadPreAccountOffer(
	fetcher: Fetcher,
	offerId: string,
	token: string,
	code: string
): Promise<PreAccountOffer> {
	const query = new URLSearchParams({ token, code });
	return readOrThrow(await fetcher(`/api/offers/${offerId}?${query.toString()}`));
}

/** Declines an Offer with the same two credentials, without joining the
 * Practice first. Accepting cannot work this way -- an attachment needs a
 * person -- but saying no must not require joining. */
export async function declinePreAccountOffer(
	fetcher: Fetcher,
	offerId: string,
	token: string,
	code: string
): Promise<{ offerId: string; state: OfferState }> {
	return readOrThrow(
		await fetcher(`/api/offers/${offerId}/decline`, {
			method: 'POST',
			headers: { 'Content-Type': 'application/json', 'X-Confirmed': 'true' },
			body: JSON.stringify({ token, code })
		})
	);
}

/** Formats an Offer's fee for display, or says so when there isn't one --
 * an employee's Offer settles her claim on the work, not her price for
 * it. The BFF omits the field entirely rather than sending a null (the
 * repo's own convention, see invoice.ts's paidAt), so undefined is the
 * only absent case to handle. */
export function formatFee(amountCents: number | undefined): string {
	if (amountCents === undefined) {
		return 'No per-Engagement fee';
	}
	return (amountCents / 100).toLocaleString('en-US', { style: 'currency', currency: 'USD' });
}

/**
 * Reports whether an Offer is still open and so still decidable.
 */
export function isOpen(offer: { state: OfferState }): boolean {
	return offer.state === 'offered';
}
